src/nvim/misc1.c

```diff
@@ -1059,3 +1059,58 @@ void add_time(char_u *buf, size_t buflen, time_t tt)
                  seconds);
   }
 }
+
+dict_T *get_v_event(save_v_event_T *sve)
+{
+  dict_T *v_event = get_vim_var_dict(VV_EVENT);
+
+  if (v_event->dv_hashtab.ht_used > 0) {
+    // recursive use of v:event, save, make empty and restore later
+    sve->sve_did_save = true;
+    sve->sve_hashtab = v_event->dv_hashtab;
+    hash_init(&v_event->dv_hashtab);
+  } else {
+    sve->sve_did_save = false;
+  }
+  return v_event;
+}
+
+void restore_v_event(dict_T *v_event, save_v_event_T *sve)
+{
+  tv_dict_free_contents(v_event);
+  if (sve->sve_did_save) {
+    v_event->dv_hashtab = sve->sve_hashtab;
+  } else {
+    hash_init(&v_event->dv_hashtab);
+  }
+}
+
+/// Fires a ModeChanged autocmd.
+void trigger_modechanged(void)
+{
+  if (!has_event(EVENT_MODECHANGED)) {
+    return;
+  }
+
+  char *mode = get_mode();
+  if (STRCMP(mode, last_mode) == 0) {
```

Comment:
This assumes that `last_mode` is never NULL because it is set elsewhere before `has_event(EVENT_MODECHANGED)`  evaluates to true. Could be a false positive in static analyzer.

Reply:
Could have used strncmp because of `MODE_MAX_LENGTH`.